cli/command/stack/kubernetes/ps.go

```diff
@@ -31,76 +37,73 @@ func RunPS(dockerCli *KubeCli, options options.PS) error {
+		return err
 	}
 
-	pods, err := fetchPods(namespace, podsClient)
+	// Fetch pods
+	if _, err := stacks.Get(stackName); err != nil {
+		return fmt.Errorf("nothing found in stack: %s", stackName)
+	}
+	pods, err := fetchPods(stackName, podsClient, filters)
 	if err != nil {
 		return err
 	}
 
 	if len(pods) == 0 {
-		return fmt.Errorf("nothing found in stack: %s", namespace)
+		return fmt.Errorf("nothing found in stack: %s", stackName)
 	}
+	return print(dockerCli, options, stackName, client, pods)
+}
 
+func print(dockerCli command.Cli, options options.PS, namespace string, client corev1.NodesGetter, pods []apiv1.Pod) error {
 	format := options.Format
-	if len(format) == 0 {
+	if format == "" {
 		format = task.DefaultFormat(dockerCli.ConfigFile(), options.Quiet)
 	}
-	nodeResolver := makeNodeResolver(options.NoResolve, client.Nodes())
 
 	tasks := make([]swarm.Task, len(pods))
 	for i, pod := range pods {
 		tasks[i] = podToTask(pod)
 	}
-	return print(dockerCli, namespace, tasks, pods, nodeResolver, !options.NoTrunc, options.Quiet, format)
-}
-
-type idResolver func(name string) (string, error)
-
-func print(dockerCli command.Cli, namespace string, tasks []swarm.Task, pods []apiv1.Pod, nodeResolver idResolver, trunc, quiet bool, format string) error {
 	sort.Stable(tasksBySlot(tasks))
 
 	names := map[string]string{}
 	nodes := map[string]string{}
 
-	tasksCtx := formatter.Context{
-		Output: dockerCli.Out(),
-		Format: formatter.NewTaskFormat(format, quiet),
-		Trunc:  trunc,
-	}
-
 	for i, task := range tasks {
-		nodeValue, err := nodeResolver(pods[i].Spec.NodeName)
+		nodeValue, err := resolveNode(options.NoResolve, client.Nodes(), pods[i].Spec.NodeName)
 		if err != nil {
 			return err
 		}
-
 		names[task.ID] = fmt.Sprintf("%s_%s", namespace, pods[i].Name)
 		nodes[task.ID] = nodeValue
 	}
 
+	tasksCtx := formatter.Context{
+		Output: dockerCli.Out(),
+		Format: formatter.NewTaskFormat(format, options.Quiet),
+		Trunc:  !options.NoTrunc,
+	}
+
 	return formatter.TaskWrite(tasksCtx, tasks, names, nodes)
 }
 
-func makeNodeResolver(noResolve bool, nodes corev1.NodeInterface) func(string) (string, error) {
+func resolveNode(noResolve bool, nodes corev1.NodeInterface, name string) (string, error) {
 	// Here we have a name and we need to resolve its identifier. To mimic swarm behavior
 	// we need to resolve the id when noresolve is set, otherwise we return the name.
 	if noResolve {
-		return func(name string) (string, error) {
-			n, err := nodes.List(metav1.ListOptions{
-				FieldSelector: fields.OneTermEqualSelector(api.ObjectNameField, name).String(),
-			})
-			if err != nil {
-				return "", err
-			}
-			if len(n.Items) != 1 {
-				return "", fmt.Errorf("could not find node '%s'", name)
-			}
-			return string(n.Items[0].UID), nil
+		n, err := nodes.List(metav1.ListOptions{
+			FieldSelector: fields.OneTermEqualSelector(api.ObjectNameField, name).String(),
+		})
+		if err != nil {
+			return "", err
+		}
+		if len(n.Items) != 1 {
```

Comment:
Could more than 1 be returned? (i.e., can node names be ambiguous?)

Reply:
In the [docs](https://kubernetes.io/docs/concepts/architecture/nodes/#management) there is `Kubernetes will create a node object internally (the representation), and validate the node by health checking based on the metadata.name field (we assume metadata.name can be resolved).`
and the example shows
 ```"metadata": {
    "name": "10.240.79.157",```
so I would assume if the name is not unique the duplicate node attempting to join the cluster will fail.

Reply:
Sorry, missed your comment: so should we skip this validation (as k8s already handles this), or keep it?